ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/ReadsKeySets.hs

```diff
@@ -0,0 +1,178 @@
+import           Ouroboros.Consensus.Storage.LedgerDB.HD.DbChangelog
+import           Ouroboros.Consensus.Storage.LedgerDB.HD.DiffSeq
+import           Ouroboros.Consensus.Storage.LedgerDB.LedgerDB (LedgerDB (..))
+
+{-------------------------------------------------------------------------------
+  Manipulate tables through the DbChangelog
+-------------------------------------------------------------------------------}
+
+data RewoundTableKeySets l =
+    RewoundTableKeySets
+      !(WithOrigin SlotNo)   -- ^ the slot to which the keys were rewounded
+      !(LedgerTables l KeysMK)
+
+rewindTableKeySets ::
+     DbChangelog l -> LedgerTables l KeysMK -> RewoundTableKeySets l
+rewindTableKeySets dblog =
+      RewoundTableKeySets
+        (changelogDiffAnchor dblog)
+
+data UnforwardedReadSets l = UnforwardedReadSets {
+    -- | The Slot number of the anchor of the 'DbChangelog' that was used when
+    -- rewinding and reading.
+    ursSeqNo  :: !(WithOrigin SlotNo)
+    -- | The values that were found in the 'BackingStore'.
+  , ursValues :: !(LedgerTables l ValuesMK)
+    -- | All the requested keys, being or not present in the 'BackingStore'.
+  , ursKeys   :: !(LedgerTables l KeysMK)
+  }
+
+forwardTableKeySets' ::
+     HasLedgerTables l
+  => WithOrigin SlotNo
+  -> LedgerTables l SeqDiffMK
+  -> UnforwardedReadSets l
+  -> Either (WithOrigin SlotNo, WithOrigin SlotNo)
+            (LedgerTables l ValuesMK)
+forwardTableKeySets' seqNo chdiffs = \(UnforwardedReadSets seqNo' values keys) ->
+    if seqNo /= seqNo' then Left (seqNo, seqNo') else
+    Right
+      $ zipLedgerTables2 forward values keys chdiffs
+  where
+    forward ::
+         (Ord k, Eq v)
+      => ValuesMK  k v
+      -> KeysMK    k v
+      -> SeqDiffMK k v
+      -> ValuesMK  k v
+    forward (ValuesMK values) (KeysMK keys) (SeqDiffMK diffs) =
+      ValuesMK $ applyDiffForKeys values keys (cumulativeDiff diffs)
+
+forwardTableKeySets ::
+     HasLedgerTables l
+  => DbChangelog l
+  -> UnforwardedReadSets l
+  -> Either (WithOrigin SlotNo, WithOrigin SlotNo)
+            (LedgerTables l ValuesMK)
+forwardTableKeySets dblog =
+  forwardTableKeySets' (changelogDiffAnchor dblog) (changelogDiffs dblog)
+
+{-------------------------------------------------------------------------------
+  The class that reads the key sets from the disk
+-------------------------------------------------------------------------------}
+
+type KeySetsReader m l = RewoundTableKeySets l -> m (UnforwardedReadSets l)
+
+readKeySets :: forall m l.
+     IOLike m
+  => LedgerBackingStore m l
+  -> KeySetsReader m l
+readKeySets (LedgerBackingStore backingStore) rew = do
+    readKeySetsVH (bsRead backingStore) rew
+
+readKeySetsVH :: forall m l.
+     Monad m
+  => (LedgerTables l KeysMK -> m (WithOrigin SlotNo, LedgerTables l ValuesMK))
+  -> RewoundTableKeySets l
+  -> m (UnforwardedReadSets l)
+readKeySetsVH readKeys (RewoundTableKeySets _seqNo rew) = do
```

Comment:
I'm assuming the `VH` prefix stands for `ValueHandle`, but we're really passing a reader function and not using any value handles here -- maybe `readKeySetsWith`?

Reply:
Done